proposed/swift-interop.md

```diff
@@ -0,0 +1,189 @@
+
+###### Error register
+
+We have selected an approach for handling the error register in the Swift calling convention:
+
+1. Use a special type named something like `SwiftError*` to indicate the error parameter
+
+This approach expresses that the to-be-called Swift function uses the Error Register in the signature and they both require signature manipulation in the JIT/AOT compilers. Like with `SwiftSelf`, we would throw an `InvalidProgramException` for a signature with multiple `SwiftError` parameters. We use a pointer-to-`SwiftError` type to indicate that the error register is a by-ref/out parameter. We don't use managed pointers as our modern JITs can reason about unmanaged pointers well enough that we do not end up losing any performance taking this route. The `UnmanagedCallersOnly` implementation will require a decent amount of JIT work to emulate a local variable for the register value, but we have prior art in the Clang implementation of the Swift error register that we can fall back on.
+
+Additionally, we have selected this design as this provides consistency with the self register and async context register handling, discussed below.
+
+We've rejected the following designs for the reasons mentioned below:
+
+1. Use an attribute like `[SwiftError]` on a `ref` or `out` parameter to indicate the error parameter.
+2. Use a special return type `SwiftReturn<T>` to indicate the error parameter and combine it with the return value.
+3. Use special helper functions `Marshal.Get/SetLastSwiftError` to get and set the last Swift error. Our various compilers will automatically emit a call to `SetLastSwiftError` from Swift functions and emit a call to `GetLastSwiftError` in the epilog of `UnmanagedCallersOnly` methods. The projection generation will emit calls to `Marshal.Get/SetLastSwiftError` to get and set these values to and from exceptions.
+4. Implicitly transform the error register into an exception at the interop boundary.
+5. Use `SwiftError` on a `ref` or `out` parameter to indicate the error parameter.
+
+We have a prototype that uses Rejected Option 1; however, using an attribute has the same limitations as the attribute option for the self register.
+
+Rejected Option 2 provides a more functional-programming-style API that is more intuitive than the `SwiftError` options above, but likely more expensive to implement. As a result, Option 2 would be better as a higher-level option and not the option implemented by the JIT/AOT compilers.
```

Comment:
> but likely more expensive to implement.

I do not think that this is measurably more expensive to implement.

Having said that, I am ok the currently selected `SwiftError*` design.